kusto/ingest/internal/queued/queued.go

```diff
@@ -105,50 +105,84 @@ func New(db, table string, mgr *resources.Manager, http *http.Client, options ..
 
 // Local ingests a local file into Kusto.
 func (i *Ingestion) Local(ctx context.Context, from string, props properties.All) error {
-	client, container, err := i.upstreamContainer()
+	containers, err := i.mgr.GetRankedStorageContainers()
 	if err != nil {
 		return err
 	}
 
-	mgrResources, err := i.mgr.Resources()
+	if len(containers) == 0 {
+		return errors.ES(
+			errors.OpFileIngest,
+			errors.KBlobstore,
+			"no Blob Storage container resources are defined, there is no container to upload to",
+		).SetNoRetry()
+	}
+
+	queues, err := i.mgr.GetRankedStorageQueues()
 	if err != nil {
 		return err
 	}
 
 	// We want to check the queue size here so we don't upload a file and then find we don't have a Kusto queue to stick
 	// it in. If we don't have a container, that is handled by containerQueue().
-	if len(mgrResources.Queues) == 0 {
+	if len(queues) == 0 {
 		return errors.ES(errors.OpFileIngest, errors.KBlobstore, "no Kusto queue resources are defined, there is no queue to upload to").SetNoRetry()
 	}
 
-	blobURL, size, err := i.localToBlob(ctx, from, client, container, &props)
-	if err != nil {
-		return err
-	}
+	// Go over all the containers and try to upload the file to each one. If we succeed, we are done.
+	for attempts, containerUri := range containers {
+		if attempts >= StorageMaxRetryPolicy {
+			return errors.ES(errors.OpFileIngest, errors.KBlobstore, "max retry policy reached").SetNoRetry()
+		}
 
-	if err := i.Blob(ctx, blobURL, size, props); err != nil {
-		return err
+		client, containerName, err := i.upstreamContainer(containerUri)
+		if err != nil {
+			i.mgr.ReportStorageResourceResult(containerUri.Account(), false)
+			continue
+		}
+
+		blobURL, size, err := i.localToBlob(ctx, from, client, containerName, &props)
+		if err == nil {
+			i.mgr.ReportStorageResourceResult(containerUri.Account(), true)
+			return i.Blob(ctx, blobURL, size, props)
+		}
+
+		// check if the error is retryable
+		if errors.Retry(err) {
```

Comment:
is checking if the error is retryable only in one place intentional?

Reply:
Yes, localToBlob(...) can throw IO errors which are not related to storage, these errors are set to NoRetry.